tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+namespace Illuminate\Tests\Database;
+
+use Illuminate\Database\Eloquent\Concerns\PreventsCircularRecursion;
+use PHPUnit\Framework\TestCase;
+
+class DatabaseConcernsPreventsCircularRecursionTest extends TestCase
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        PreventsCircularRecursionWithRecursiveMethod::$globalStack = 0;
+    }
+
+    public function testRecursiveCallsArePreventedWithoutPreventingSubsequentCalls()
+    {
+        $instance = new PreventsCircularRecursionWithRecursiveMethod();
+
+        $this->assertEquals(0, PreventsCircularRecursionWithRecursiveMethod::$globalStack);
+        $this->assertEquals(0, $instance->instanceStack);
+
+        $this->assertEquals(0, $instance->callStack());
```

Comment:
Why is this zero?

On the first iteration, the callback is executed and `$this->instanceStack` becomes `1` and returns the recursive call.

On the recursive call, the recursion is guarded, then it returns `$this->instanceStack`, which should be `1`, right?

Reply:
The "default" value that gets returned when it recurses is `$this->instanceStack` as it was passed in when it was originally called (i.e. `0`). If the default was a callback, it would resolve later and return `1`.

So in essence, the first call does this:
```php
public function callStack(): int
    return $this->withoutRecursion(
        function () {
            static::$globalStack++;
            $this->instanceStack++;

             return $this->callStack(); // gets the cached/default value that was passed in below
        },
        $this->instanceStack // currently 0
    );
}
```

If the call was made using a callback method that resolve _later_: 
```php
public function callStack(): int
    return $this->withoutRecursion(
        function () {
            static::$globalStack++;
            $this->instanceStack++;

             return $this->callStack(); // calls the callback which gets the current version now (which is 1), then caches that value for subsequent calls
        },
        fn () => $this->instanceStack
    );
}
```

Reply:
@rodrigopedra thanks for reminding me to add some additional tests for the "default callback" feature

Reply:
I get what you mean, it behaves as it is avoiding the recursion altogether, right?

But either way, it still iterates the recursion a second time, but instead of returning the second call's `$default` value, it pulls the cached `$default` value from the first call, right?

I mean:

```php
$instance->callStack(); // first time

// Implementation calls:
$this->withoutRecursion(fn () => ..., 0);

// Inside the callback it calls... 
$instance->callStack(); // ...a second time time
// from the return statement, which is a recursive call

// Implementation calls this, again...
$this->withoutRecursion(fn () => ..., 1);
// ...where it detects a recursive call, ignores the current default
// and pulls the original default from the cache from the first call
```

IMO, I find it a bit odd, from my gut I'd expect the `$default` to be used when the `$callback` can't be used anymore, which would be the `$default` from the second call.

But I see what you mean, it is more like, "when we can't recurse anymore, use this default instead", and then you use the cached one.

I guess it is just different point of views.

Thanks for your patience!

Reply:
Yeah, there's potentially a different or "better" name for that variable, but nothing that sprang to mind as being particularly "Laravel-ish". Maybe `$onRecursion`, or `$recurseWith`, or `$andThen`, but nothing that really _stands out_.